aten/src/ATen/native/Convolution.cpp

```diff
@@ -364,10 +377,19 @@ at::Tensor _convolution(
     bool transposed_, IntArrayRef output_padding_, int64_t groups_,
     bool benchmark, bool deterministic, bool cudnn_enabled) {
 
-  auto input = input_r.contiguous();
+  const bool input_is_mkldnn = (input_r.type_id() == MkldnnCPUTensorId());
+  auto input = input_r;
+  if (!input_is_mkldnn) {
+    input = input.contiguous();
+  }
   auto weight = weight_r;
```

Comment:
shall we provide MKLDNN tensor with `contiguous()`?
let `is_contiguous()` return `true` will do the job? In case we have a MKLDNN tensor, it should always be contiguous.

Reply:
The question is how we can define sound semantics of `contiguous` for opaque tensors. It does not make sense to have `is_contiguous` return true when `contiguous` is not well defined.

Reply:
you can check `TensorTypeId` of course, i did not see any discomfortable to let `MkldnnCPUTensorId` return `true` for `tensor.is_contiguous()`.

Reply:
I'm a little worried of making contiguous work on MKLDNN as the semantics are not well-defined. Maybe add a wrapper function `contiguous_or_mkldnn` and call it only in this implementation?

Reply:
@dzhulgakov Indeed the semantics indicates that `mkldnn` blocked layout should be non-contiguous while at the same time mkldnn function calls just need to know that whether the buffer is a physical contiguous chunk of memory (which is always yes for `mkldnn` tensor).

In case we don't inherit contiguous() for `mkldnn` tensor. Another simple method will be move all `input.contiguous()`, `weight.contiguous()`, `bias.contiguous()` from `_convolution()` to each underlying device implementations: a.k.a `cudnn_convolution()` , `mkldnn_convolution()` etc.
In this case, at the upper level from `_convolution()` will be unified (look cleaner) and inside `mkldnn_convolution` we are free to do something like:
```C
// at::native::mkldnn_convolution()
input = is_input_mkldnn ? input : input.contiguous()
weight = is_weight_mkldnn ? weight : weight.contiguous()
```

and we can do this kind of check ONLY inside mkldnn function calls, so `mkldnn` tensor don't have to inherit `contiguous()`